Fix the returned cumulative gas used and log index in transaction receipts

ngotchac opened the pull request:
<!--
Thank you for using Hardhat and taking the time to send a Pull Request!

Please consider the checklist items below, keep the ones that make sense for your PR, and delete the ones that don't.  If you are introducing a new feature, please discuss it in an Issue or with someone from the team before submitting your change.
-->

- [x] (this PR includes a **bug fix**) Relevant tests have been included.

---
It seems that the cumulativeGasUsed and logIndex fields in transaction receipts were not correct.

For cumulativeGasUsed: the `gas` field of the receipt is actually already the `cumulativeGasUsed` ; so we can use it directly,
or compute it from the `gasUsed` value of the result.

For the logIndex, it actually isn't per transaction, but within a block.

Note : how `logIndex` should be set is not actually in the specs, but that's how it's commonly implemented, cf. [Geth implementation](https://github.com/ethereum/go-ethereum/blob/87a11a87c2bcbeedce9e088698b48d8fa2a09450/core/types/receipt.go#L272-L314)

fvictorio commented:
This is great @ngotchac, thanks a lot!

LGTM, I only refactored the test so that it's more consistent with how we usually write them.

alcuadrado commented on packages/hardhat-core/test/internal/hardhat-network/provider/modules/eth/methods/getTransactionReceipt.ts:
```diff
@@ -51,6 +54,63 @@ describe("Eth module", function () {
+            [txHash2]
+          );
+
+          assert.equal(rpcQuantityToNumber(receipt1.gasUsed), 44945);
```
this is a bit nitpicky, but i think checking the exact number of gas here will lead to this tests breaking in the future (e.g. after a new hardfork). I'd test that these pseudocode conditions hold:

```js
receipt1.gasUsed === receipt1.cumulativeGasUsed
receipt2.cumulativeGasUsed === receipt1.cumulativeGasUsed + receipt2.gasUsed
```

fvictorio replied:
That's actually one of the things I modified from @ngotchac original code. I did it like this because I like to have the less amount of logic possible in a test, even if that means having to update some values after a harfork. But I'll change it back, no problem.

alcuadrado commented:
Thanks for this PR, @ngotchac ! Good catches.

I left a minor comment, and also restarted the CI. 

@fvictorio have you seen another OOM error in the CI? I think this is a first one.

fvictorio commented:
> @fvictorio have you seen another OOM error in the CI? I think this is a first one.

Yes, [here](https://github.com/nomiclabs/hardhat/pull/1556#issuecomment-863144156)

ngotchac commented:
Hey, just to let you know that I rebased this PR, so it should now be merge-able.

ngotchac commented:
Hey @fvictorio and @alcuadrado ; I just found out that I didn't completely fix the issue on my first attempt, since I missed the logs returned by a forked node where the log index was completely off. So I just fixed it, and I updated a test to make sure it works.
I also took the liberty to replace the tests mentioning [0x5acc84a3e955bdd76467d3348077d003f00ffb97](https://etherscan.io/address/0x5acc84a3e955bdd76467d3348077d003f00ffb97) since it's marked as _Ponzi_ on Etherscan. I hope that's OK

fvictorio commented:
> I also took the liberty to replace the tests mentioning [0x5acc84a3e955bdd76467d3348077d003f00ffb97](https://etherscan.io/address/0x5acc84a3e955bdd76467d3348077d003f00ffb97) since it's marked as _Ponzi_ on Etherscan. I hope that's OK

Ugh, thanks for that, it's definitely ok!

And thanks for updating the PR. We are a bit behind (well, more than a bit...) on reviewing open pull requests. Sorry about that!

ngotchac commented:
No worries, I'm glad I could make the latest release ;)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/nomiclabs/hardhat?pullRequest=1560) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/nomiclabs/hardhat?pullRequest=1560) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/nomiclabs/hardhat?pullRequest=1560) it.</sub>
